lib/SILOptimizer/Transforms/RetainReleaseCodeMotion.cpp

```diff
@@ -0,0 +1,1070 @@
+#include "swift/SILOptimizer/Utils/Local.h"
+#include "llvm/ADT/BitVector.h"
+#include "llvm/ADT/DenseSet.h"
+#include "llvm/ADT/Statistic.h"
+#include "llvm/Support/Allocator.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Debug.h"
+
+using namespace swift;
+
+STATISTIC(NumRetainsSunk, "Number of retains sunk");
+STATISTIC(NumReleasesHoisted, "Number of releases hoisted");
+
+//===----------------------------------------------------------------------===//
+//                             Utility 
+//===----------------------------------------------------------------------===//
+
+static bool isRetainInstruction(SILInstruction *I) {
+  return isa<StrongRetainInst>(I) || isa<RetainValueInst>(I);
+}
+
+static bool isReleaseInstruction(SILInstruction *I) {
+  return isa<StrongReleaseInst>(I) || isa<ReleaseValueInst>(I);
+}
+
+/// Creates an increment on \p Ptr before insertion point \p InsertPt that
+/// creates a strong_retain if \p Ptr has reference semantics itself or a
+/// retain_value if \p Ptr is a non-trivial value without reference-semantics.
+static SILInstruction *
+createIncrementBefore(SILValue Ptr, SILInstruction *InsertPt) {
+  // Set up the builder we use to insert at our insertion point.
+  SILBuilder B(InsertPt);
+  auto Loc = RegularLocation(SourceLoc());
+
+  // If Ptr is refcounted itself, create the strong_retain and
+  // return.
+  if (Ptr->getType().isReferenceCounted(B.getModule()))
+    return B.createStrongRetain(Loc, Ptr, Atomicity::Atomic);
+
+  // Otherwise, create the retain_value.
+  return B.createRetainValue(Loc, Ptr, Atomicity::Atomic);
+}
+
+/// Creates a decrement on \p Ptr before insertion point \p InsertPt that
+/// creates a strong_release if \p Ptr has reference semantics itself or
+/// a release_value if \p Ptr is a non-trivial value without reference-semantics.
+static SILInstruction *
+createDecrementBefore(SILValue Ptr, SILInstruction *InsertPt) {
+  // Setup the builder we will use to insert at our insertion point.
+  SILBuilder B(InsertPt);
+  auto Loc = RegularLocation(SourceLoc());
+
+  // If Ptr has reference semantics itself, create a strong_release.
+  if (Ptr->getType().isReferenceCounted(B.getModule()))
+    return B.createStrongRelease(Loc, Ptr, Atomicity::Atomic);
+
+  // Otherwise create a release value.
+  return B.createReleaseValue(Loc, Ptr, Atomicity::Atomic);
+}
+
+//===----------------------------------------------------------------------===//
+//                             Block State 
+//===----------------------------------------------------------------------===//
+
+struct BlockState {
+  /// A bit vector for which the ith bit represents the ith refcounted root in
+  /// RCRootVault.
+  ///
+  /// NOTE: we could do the data flow with BBSetIn or BBSetOut, but that would
+  /// require us to create a temporary copy to check whether the BBSet has
+  /// changed after the genset and killset has been applied.
+  llvm::SmallBitVector BBSetIn;
```

Comment:
Have you measured the cardinality of any of these sets? For example getting a histogram of how large these tend to be at some point(s) in solving the equations, compared to the allocation size, could be interesting. I suspect these are relatively small so it might not matter, but in some cases sparse bit vectors can be a large win over dense ones both in terms of memory consumed and work involved in intersections and unions.